source/MapDetailPanel.cpp

```diff
@@ -573,13 +573,33 @@ void MapDetailPanel::DrawKey()
 		// four largest visible governments are labeled in the legend.
 		vector<pair<double, const Government *>> distances;
 		for(const auto &it : closeGovernments)
+		{
+			if(!it.first)
+				continue;
 			distances.emplace_back(it.second, it.first);
+		}
 		sort(distances.begin(), distances.end());
-		for(unsigned i = 0; i < 4 && i < distances.size(); ++i)
+		int drawn = 0;
+		multimap<const string, const Color *> alreadyDisplayed;
+		for(auto &it : distances)
 		{
-			RingShader::Draw(pos, OUTER, INNER, GovernmentColor(distances[i].second));
-			font.Draw(distances[i].second->GetName(), pos + textOff, dim);
+			const string &displayName = it.second->GetName();
+			const Color &displayColor = GovernmentColor(it.second);
+			auto lower = alreadyDisplayed.lower_bound(displayName);
+			auto upper = alreadyDisplayed.upper_bound(displayName);
+			auto found = find_if(lower, upper, [&displayColor](const pair<const string, const Color *> &item)
+					{
+						return *item.second == displayColor;
+					});
+			if(found != upper)
+				continue;
+			RingShader::Draw(pos, OUTER, INNER, GovernmentColor(it.second));
+			font.Draw(it.second->GetName(), pos + textOff, dim);
 			pos.Y() += 20.;
+			alreadyDisplayed.emplace(displayName, &displayColor);
```

Comment:
You're storing a pointer to a variable that will get destroyed at the end of the current iteration.

Reply:
That variable is supposed to be a reference.

Reply:
`GovernmentColor` returns a temporary, so the temporary gets lifetime extended and gets destroyed when the reference does.

Reply:
`Government::GetColor()`?

Reply:
yeah that works